### PR TITLE
Added method in python runner to get eventNumber for TC-SMOKECO tests

### LINKED ISSUE #27539: [Feature] Get the latest eventNumber in the chip-tool test
### Feature description

There is a possible solution to get the event number so that I can read the specified event, but it requires adding the Unit Testing cluster.

```sh
- label: "Generate an event on the accessory"
  cluster: "Unit Testing"
  endpoint: 0
  command: "TestEmitTestEventRequest"
  arguments:
      values:
          - name: "arg1"
            value: 1
          - name: "arg2"
            value: 2
          - name: "arg3"
            value: true
  response:
      values:
          - name: "value"
            saveAs: eventNumber
```

Is it possible to add a simpler command to get the latest eventNumber?

Or add a `readLastEvent` command.

### Platform

other

### Platform Version(s)

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by ericzijian1994**
Proposing the method to fixes #27539

- This PR tried to -

Adding a method in Python runner to get eventNumber in order to get the TC-SMOKECO test to work, I understand this could not be a good implementation and may need improvement/use another alternative. 

- The problem was -

The `readEvent` command from eventTest returns a list of events. But in yaml, the value I expect needs to match the value actually returned. And the eventNumber parameter in yaml acts as a filter that allows me to accomplish this. 
If I know the event number before a readEvent step, I can get the latest one located and read the event generated after this eventNumber to pass the test plan, which also makes it easier for automation.

- Why such command is essential -
In the testing of the smoke and CO Test plan, we will verify a series of new events that are received. The EventNumber is used to distinguish whether an event is the most recent one. However, the issue lies in the fact that the EventNumber provided in the testEvents currently relies on an initial eventNumber in a configuration file. (I have set it to 0 in the config, which results in retrieving all records.)
Since I didn't find certain commands only returned the latest event sequence number. I can't do the effective update to flag the new starting point with the EventNumber to search event. 
The readEvent returns a list of all corresponding events found. Due to the fact that the test plan will test multiple times for different scenarios triggering the same event, this leads to the continuous growth of the returned list... Consequently, if testing persists, the returned records will become unpredictable and lead to test failure.

Example test Ymal:
```
config:
......
    EVENT_NUMBER:
        type: int64u
        defaultValue: 0

label: "TH reads LowBattery event from DUT"
      PICS: SMOKECO.S.E02
      command: "readEvent"
      event: "LowBattery"
      EVENT_NUMBER: EVENT_NUMBER + 1
      response:
          value: { AlarmSeverityLevel: 1 }
....
label: "TH reads LowBattery event from DUT"
      PICS: SMOKECO.S.E02
      command: "readEvent"
      event: "LowBattery"
      EVENT_NUMBER: EVENT_NUMBER + 2
      response:
          value: { AlarmSeverityLevel: 2 } //Here could have two values returned. 
```

Until then, I've tried to modify EVENT_NUMBER at config and letting testers enter it manually.
But there are some issues with using the eventNumber parameter in TC-SMOKECO:

1. The TH currently only accepts the endpoint parameter, setting EVENT_NUMBER manually may fail/not take effect. Even if it works, it can be troublesome for testing. Such manual operation requires finding the value manually first before executing each TC, and then needing to modify the config value after each TC. So they will have a relatively clean list from the readEvent for individual TC, which will make the tester very inconvenient and also not guarantee some accidentally generated event will not interfere with the readEvent verification.
2. EVENT_NUMBER +1 in each readEvent step, if any test item is skipped due to PICS, it will also cause failure at the test.

- To make TC work -

So this would need a new method to get the latest event automatically.
My logic for the change is to 

1. Add the `last_event_number` variable to the python runner to record the latest event number of the readEvent.
2. At the beginning of each yaml, add a `readEvent` with no response to let the last_event_number have a value.
3. In the subsequent readEvent step, if the `isLastEventNumber` parameter is true, the eventNumber parameter is set to last_event_number. 

This will make a method for tracking the last event number. By running tests, this method worked with the PR proposed here.

### CODE REVIEW

File: src/app/ReadClient.cpp [hunk 1]
```diff
@@ -1192,6 +1203,10 @@ CHIP_ERROR ReadClient::GetMinEventNumber(const ReadPrepareParams & aReadPrepareP
             // We want to start with the first event _after_ the last one we received.
             aEventMin.SetValue(aEventMin.Value() + 1);
         }
+        else if (aReadPrepareParams.mIsLastEventNumber.HasValue() && aReadPrepareParams.mIsLastEventNumber.Value())
```
Comment (yunhanw-google) [inline src/app/ReadClient.cpp:1206]: Please use cluster state cache in chip-tool, where it has recorded the latest event number, or mimic cluster state cache to update the existing callback in chip-tool and record the latest event number.
If you see line above, we try to load the latest event number from callback, and start eventnumber + 1 in next try. 
        ReturnErrorOnFailure(mpCallback.GetHighestReceivedEventNumber(aEventMin));
        if (aEventMin.HasValue())
        {
            // We want to start with the first event _after_ the last one we received.
            aEventMin.SetValue(aEventMin.Value() + 1);
        }
Please don't modify the ReadClient here, thanks

**Review by yunhanw-google — CHANGES_REQUESTED** [review]

File: src/app/ReadClient.cpp (hunk 1 above)
Comment (ericzijian1994) [reply src/app/ReadClient.cpp:1206]: Thanks for the review, I did notice GetHighestReceivedEventNumber but it actually functions as ClearValue. And cluster state cache doesn't seem to work for posix chip-tool, I don't know how to modify it, could you provide some guidance on that, thanks!

Comment (yunhanw-google) [reply src/app/ReadClient.cpp:1206]: https://github.com/project-chip/connectedhomeip/blob/98bdcd946fb79ccda232335b824cbce6035a53d5/src/app/ClusterStateCache.cpp#L198
the mHighestReceivedEventNumber is updated with the latest number here when we received the event from report, so it is not clearValue, then ReadClient would use next event number to construct event filter based upon  mHighestReceivedEventNumber. one possible integration is to inherit cluster state cache from https://github.com/project-chip/connectedhomeip/blob/98bdcd946fb79ccda232335b824cbce6035a53d5/examples/chip-tool/commands/clusters/ReportCommand.h#L26, this is one reference code in android jni https://github.com/project-chip/connectedhomeip/blob/98bdcd946fb79ccda232335b824cbce6035a53d5/src/controller/java/AndroidCallbacks.h#L85, you could also find the code usage in ios and test code
posix chip-tool support c++ stl, technically it should work when integrating with ClusterStateCache

Comment (ericzijian1994) [reply src/app/ReadClient.cpp:1206]: Thanks for the reply, but it seems that this is going to replace the callback, which might be a big change. I've heard that Python runner will replace c++ runner, is this true? Maybe I can remove this part of the change for now, can you review my other changes, thanks!

Comment (yunhanw-google) [reply src/app/ReadClient.cpp:1206]: @vivien-apple maybe we should consider to integrate cluster state cache into callback for chip-tool this time, currently android/ios controller both has integrated cluster state cache.

Comment (ericzijian1994) [reply src/app/ReadClient.cpp]: I removed the C++ runner changes for now to avoid conflicts with vivien's PR.

Comment (ericzijian1994) [reply src/app/ReadClient.cpp]: If it is necessary to add, I can create a new PR.

Comment (ericzijian1994) [reply src/app/ReadClient.cpp]: I've removed this part of the change anyway to avoid conflict with vivien's change. If this is really needed, I'll create a new PR.

File: src/app/tests/suites/certification/Test_TC_SMOKECO_1_1.yaml [hunk 2]
```diff
@@ -23,23 +23,23 @@ config:
     endpoint: 1
 
 tests:
-    - label: "Commission DUT to TH"
+    - label: "Step 1: Commission DUT to TH"
```
Comment (andy31415) [inline src/app/tests/suites/certification/Test_TC_SMOKECO_1_1.yaml:26]: a lot of these changes are not "add method to get eventNumber". Please split them in separate PRs so that we can review them individually.

**Review by andy31415 — CHANGES_REQUESTED** [review]
Title says "added method in chip-tool to get eventNumber" yet no chip-tool changes contained in the PR.

Please update accordingly:
  - split changes in independent bits (yaml text edits, test disabling, event number processing)
  - update description on what each PR does, how and why.

**Review by andy31415 — CHANGES_REQUESTED** [review]
I do not believe reading logs is a sustainable approach. Our runners have access to the entire stack in theory, so it should be able to install appropriate callbacks or subscriptions to receive events and keep track of them.

File: scripts/py_matter_yamltests/matter_yamltests/runner.py [hunk 3]
```diff
@@ -175,6 +180,9 @@ async def _run(self, parser: TestParser, config: TestRunnerConfig):
 
             test_duration = 0
             for idx, request in enumerate(parser.tests):
+                if request.is_event and request.is_last_event_number:
```
Comment (andy31415) [inline scripts/py_matter_yamltests/matter_yamltests/runner.py:183]: `is_last_event_number` is an odd name. I expect "is" to mean a property of something, where as this seems to control the request itself. maybe "use_last_event_number" or "new_events_only" or something like that.

File: src/app/tests/suites/certification/Test_TC_SMOKECO_2_2.yaml [hunk 4]
```diff
@@ -121,7 +123,7 @@ tests:
       PICS: SMOKECO.S.E00
       command: "readEvent"
       event: "SmokeAlarm"
-      EVENT_NUMBER: EVENT_NUMBER + 1
+      isLastEventNumber: true
```
Comment (andy31415) [inline src/app/tests/suites/certification/Test_TC_SMOKECO_2_2.yaml:126]: use a different name to specify what we do. like one of:

```
skipOlderEvents: true
fromLastEventNumber: true
event_number: "useLastEvent"  # with special handling in this case
```

Comment (andy31415) [reply src/app/tests/suites/certification/Test_TC_SMOKECO_2_2.yaml:126]: Or we can have the last event number as an implicit variable set during post process and then we can have:

```
EVENT_NUMBER: "LATEST_EVENT_NUMBER"
```

since we store that every time anyway.

File: scripts/py_matter_yamltests/matter_yamltests/runner.py (hunk 3 above)
Comment (ericzijian1994) [reply scripts/py_matter_yamltests/matter_yamltests/runner.py]: I think new_events_only is more appropriate, because it may actually have more than one.

**Review by yunhanw-google — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]
Approving. I believe as a followup we should not have a separate per runner variable and instead we should use the `variable storage` to keep track of the last event number.
